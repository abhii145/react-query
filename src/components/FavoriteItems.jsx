import { useDispatch, useSelector } from "react-redux"
import { FaTrash } from "react-icons/fa6"
import { removeFromCart } from "../store/cartSlice"
import { toggleFavorite } from "../store/favoritesSlice"

const FavoriteItems = () => {
  const dispatch = useDispatch()
    const favoriteItems = useSelector((state) => state.favorites.favorites)
    console.log(favoriteItems)

    const handleRemoveFromFavorites = (item) => {
      dispatch(toggleFavorite(item))
    }

  return (
    <div className="font-sans max-w-4xl mx-auto p-4">
      <h1 className="text-2xl font-extrabold text-gray-800">Your Favorites</h1>
      <div className="grid md:grid-cols-3 gap-4 mt-8">
        {favoriteItems.length > 0 ? (
          favoriteItems.map((item) => (
            <div
              key={item.id}
              className="flex gap-4 bg-white px-4 py-6 rounded-md shadow-[0_2px_12px_-3px_rgba(6,81,237,0.3)]"
            >
              <div className="w-28 h-28 max-sm:w-24 max-sm:h-24 shrink-0">
                <img
                  src={item.thumbnail}
                  className="w-full h-full object-contain"
                />
              </div>
              <div className="flex flex-col gap-4">
                <h3 className="text-base font-bold text-gray-800">
                  {item.name}
                </h3>
                <p className="text-sm font-semibold text-gray-500 mt-2">
                  {item.title}
                </p>
              </div>
              <div className="ml-auto flex flex-col">
                <FaTrash
                  size={20}
                  color="red"
                  onClick={() => handleRemoveFromFavorites(item)}
                  className="cursor-pointer"
                />
              </div>
            </div>
          ))
        ) : (
          <p>No favorite items yet</p>
        )}
      </div>
    </div>
  )
}

export default FavoriteItems
