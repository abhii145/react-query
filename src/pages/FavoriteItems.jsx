import { useDispatch, useSelector } from "react-redux"
import { FaHeart, FaTrash } from "react-icons/fa6"
import { toggleFavorite } from "../store/favoritesSlice"
import { Link } from "react-router"
import EmptyStateCard from "../components/EmptyStateCard"
import { toast } from "react-toastify"

const FavoriteItems = () => {
  const dispatch = useDispatch()
  const favoriteItems = useSelector((state) => state.favorites.favorites)

  const handleRemoveFromFavorites = (item) => {
    dispatch(toggleFavorite(item))
    toast.error("Removed from favorites", {
      position: "bottom-right",
      autoClose: 800,
    })
  }

  return (
    <div className="bg-gray-100">
      <div className="mx-auto max-w-2xl px-4 py-8 sm:px-6 sm:py-12 lg:max-w-7xl lg:px-8">
        <h1 className="text-3xl font-extrabold text-gray-900 mb-6">
          Your Favorites
        </h1>
        {favoriteItems.length === 0 ? (
          <EmptyStateCard
            icon={FaHeart}
            heading="Your Favorites is Empty"
            subheading="Start adding your favorite items now."
            buttonLabel="Shop Now"
          />
        ) : (
          <div className="mt-6 grid grid-cols-1 gap-x-6 gap-y-10 sm:grid-cols-2 lg:grid-cols-4 xl:gap-x-8">
            {favoriteItems.map((item) => (
              <div
                key={item.id}
                className="group relative bg-gray-300 rounded-md overflow-hidden"
              >
                {/* Product Image */}
                <img
                  loading="lazy"
                  alt={item.title}
                  src={item.thumbnail}
                  className="aspect-square w-full object-cover group-hover:opacity-75"
                />
                {/* Title and Price Overlay */}
                <div className="absolute inset-x-0 bottom-0 bg-black/70 text-white p-4">
                  <div className="flex justify-between items-center">
                    <h3 className="text-sm font-bold">
                      <Link to={`/product/${item.id}`}>
                        <span aria-hidden="true" className="absolute inset-0" />
                        {item.name}
                      </Link>
                    </h3>
                    <p className="text-sm font-bold">${item.price}</p>
                  </div>
                  <p className="mt-1 text-sm">{item.title}</p>
                </div>
                {/* Delete Button */}
                <div className="absolute top-2 right-2">
                  <button
                    onClick={() => handleRemoveFromFavorites(item)}
                    className="p-2 bg-red-200 rounded-full hover:bg-red-300"
                    data-testid={`delete-${item.id}`}
                  >
                    <FaTrash size={16} color="red" />
                  </button>
                </div>
              </div>
            ))}
          </div>
        )}
      </div>
    </div>
  )
}

export default FavoriteItems
