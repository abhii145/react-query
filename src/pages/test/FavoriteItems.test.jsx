import { render, screen, fireEvent } from "@testing-library/react"
import { Provider } from "react-redux"
import { configureStore } from "@reduxjs/toolkit"
import { MemoryRouter } from "react-router-dom"
import favoritesReducer from "../../store/favoritesSlice"
import FavoriteItems from "../FavoriteItems"
import { toast } from "react-toastify"
import { expect, describe, vi, beforeEach, it } from "vitest"

vi.mock("react-toastify", () => ({
  toast: {
    error: vi.fn(),
    success: vi.fn(),
  },
}))

const createMockStore = (initialState) => {
  const store = configureStore({
    reducer: {
      favorites: favoritesReducer,
    },
    preloadedState: initialState,
  })
  store.dispatch = vi.fn(store.dispatch)
  return store
}

describe("FavoriteItems", () => {
  let store

  beforeEach(() => {
    store = createMockStore({
      favorites: {
        favorites: [
          {
            id: 1,
            title: "Favorite Product 1",
            name: "Product 1",
            price: 20,
            thumbnail: "thumbnail1.jpg",
          },
        ],
      },
    })
  })


    it("matches snapshot", () => {
      const { asFragment } = render(
        <Provider store={store}>
          <MemoryRouter>
            <FavoriteItems />
          </MemoryRouter>
        </Provider>
      )
      expect(asFragment()).toMatchSnapshot()
    })

  it("renders empty favorites message when there are no favorite items", () => {
    store = createMockStore({
      favorites: { favorites: [] },
    })

    render(
      <Provider store={store}>
        <MemoryRouter>
          <FavoriteItems />
        </MemoryRouter>
      </Provider>
    )

    expect(screen.getByText("Your Favorites is Empty")).toBeInTheDocument()
  })

  it("renders favorite items correctly", () => {
    render(
      <Provider store={store}>
        <MemoryRouter>
          <FavoriteItems />
        </MemoryRouter>
      </Provider>
    )

    expect(screen.getByText("Favorite Product 1")).toBeInTheDocument()
    expect(screen.getByText("$20")).toBeInTheDocument()
  })

  it("handles remove from favorites", () => {
    render(
      <Provider store={store}>
        <MemoryRouter>
          <FavoriteItems />
        </MemoryRouter>
      </Provider>
    )

    fireEvent.click(screen.getByTestId("delete-1"))
    const actions = store.dispatch.mock.calls.map((call) => call[0])
    expect(actions).toContainEqual({
      type: "favorites/toggleFavorite",
      payload: {
        id: 1,
        title: "Favorite Product 1",
        name: "Product 1",
        price: 20,
        thumbnail: "thumbnail1.jpg",
      },
    })
    expect(toast.error).toHaveBeenCalledWith("Removed from favorites", {
      position: "bottom-right",
      autoClose: 800,
    })
  })


})