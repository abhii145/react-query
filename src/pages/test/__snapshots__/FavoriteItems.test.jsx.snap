// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FavoriteItems > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="bg-gray-100"
  >
    <div
      class="mx-auto max-w-2xl px-4 py-8 sm:px-6 sm:py-12 lg:max-w-7xl lg:px-8"
    >
      <h1
        class="text-3xl font-extrabold text-gray-900 mb-6"
      >
        Your Favorites
      </h1>
      <div
        class="mt-6 grid grid-cols-1 gap-x-6 gap-y-10 sm:grid-cols-2 lg:grid-cols-4 xl:gap-x-8"
      >
        <div
          class="group relative bg-gray-300 rounded-md overflow-hidden"
        >
          <img
            alt="Favorite Product 1"
            class="aspect-square w-full object-cover group-hover:opacity-75"
            loading="lazy"
            src="thumbnail1.jpg"
          />
          <div
            class="absolute inset-x-0 bottom-0 bg-black/70 text-white p-4"
          >
            <div
              class="flex justify-between items-center"
            >
              <h3
                class="text-sm font-bold"
              >
                <a
                  data-discover="true"
                  href="/product/1"
                >
                  <span
                    aria-hidden="true"
                    class="absolute inset-0"
                  />
                  Product 1
                </a>
              </h3>
              <p
                class="text-sm font-bold"
              >
                $20
              </p>
            </div>
            <p
              class="mt-1 text-sm"
            >
              Favorite Product 1
            </p>
          </div>
          <div
            class="absolute top-2 right-2"
          >
            <button
              class="p-2 bg-red-200 rounded-full hover:bg-red-300"
              data-testid="delete-1"
            >
              <svg
                color="red"
                fill="currentColor"
                height="16"
                stroke="currentColor"
                stroke-width="0"
                style="color: red;"
                viewBox="0 0 448 512"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M135.2 17.7L128 32H32C14.3 32 0 46.3 0 64S14.3 96 32 96H416c17.7 0 32-14.3 32-32s-14.3-32-32-32H320l-7.2-14.3C307.4 6.8 296.3 0 284.2 0H163.8c-12.1 0-23.2 6.8-28.6 17.7zM416 128H32L53.2 467c1.6 25.3 22.6 45 47.9 45H346.9c25.3 0 46.3-19.7 47.9-45L416 128z"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
